### PR TITLE
Game Lobby Settings 

### PR DESCRIPTION
**Pull request by yds05238**
## Overview

Added game settings to the game lobby (Stardust Spawn Rate, Planet Gravity Strength, Num of Stardust Colors, and Planet condition for Game Win). Updated game scene to handle settings passed in from the menu. Switched to a smaller resized outerLayerAnimation texture since the original file was too large for the mobile simulator (I can switch back to the original file if this causes any issues). Quick fix on mod 0 issue when initializing spawnRand in addStardust method (also let me know if this causes any issues). 

## Changes Made

- Added widgets to menu.json for game lobby setting inputs/labels and fixed ninepatch sizing errors in menu assets
- Updated MenuScene and LobbyMenu to handle game settings 
- Changed LobbyMenu's update and setDisplay methods for Host/Client views
- Added fields for stardust spawn rate and available color count to GameScene 
- Added win condition (game length) and planet gravity strength to PlanetModel 
- Updated CollisionController and OpponentPlanet to use the setting values

## Next Steps

Need to update the game lobby to match the designs when the designs are finalized. 

## Screenshots

- Host Game Lobby 

gamelobby_host

- Client Game Lobby

gamelobby_client

### CODE REVIEW

File: source/CIGameUpdate.h [hunk 1]
```diff
@@ -135,13 +135,6 @@ class GameUpdate {
     int getTimestamp() {
         return _timestamp;
     }
-    
-    /**
-     * Returns whether the player sending the game update has won.
-     */
-    bool didPlayerWin() {
-        return _planet->isWinner();
-    }
```
Comment (henryz2000) [inline source/CIGameUpdate.h:144]: why did we get rid of this?

File: source/CILobbyMenu.cpp [hunk 2]
```diff
@@ -105,45 +206,116 @@ void LobbyMenu::setDisplay(bool onDisplay) {
...
+            
+            state = _nextState = MenuState::GameLobby;
+            break;
+        }
         case MenuState::JoinToLobby:
-            // handle transitioning into GameLobby
-            setDisplay(true);
+        {
+            // handle displaying for Clients
+            _isClient = true;
```
Comment (henryz2000) [inline source/CILobbyMenu.cpp:255]: why did we get rid of `isHost` just to have `isClient`?

File: source/CIMenuScene.h [hunk 3]
```diff
@@ -180,9 +205,70 @@ class MenuScene : public cugl::Scene2 {
...
+
+    /**
+     * Returns the player settings value. The player settings value contains
+     * 4 fields (PlayerName, Volume, MusicOn, ParallaxOn).
+     *
+     * The method is used to retrieve and save the player settings.
+     *
+     * @param playerSettings reference to the json value for player settings from App
+     */
+    void getPlayerSettings(std::shared_ptr<cugl::JsonValue>& playerSettings) {
```
Comment (henryz2000) [inline source/CIMenuScene.h:220]: maybe `appendPlayerSettings` instead?

**Review by henryz2000 — COMMENTED** [review]
maybe explain what game length is or change the title to something more descriptive and precise

File: source/CIGameUpdate.h (hunk 1 above)
Comment (yds05238) [reply source/CIGameUpdate.h:144]: oh it seemed like we weren't using it and was overlapping with the planetmodel's isWinner method. Were we planning on using it somewhere? If so I can add it back in.

File: source/CIApp.cpp [hunk 4]
```diff
@@ -74,14 +74,11 @@ void CoreImpactApp::onStartup() {
  */
 void CoreImpactApp::onShutdown() {
     // save settings file
-    std::shared_ptr<cugl::JsonWriter> _writer = JsonWriter::alloc(Application::getSaveDirectory().append("settings.json"));
-    std::shared_ptr<cugl::JsonValue> _settings = JsonValue::allocObject();
-    _settings->appendValue("PlayerName", _menu.getPlayerName());
-    _settings->appendValue("Volume", _menu.getVolume());
-    _settings->appendValue("MusicOn", _menu.isMusicOn());
-    _settings->appendValue("ParallaxOn", _menu.isParallaxOn());
+    auto _writer = JsonWriter::alloc(Application::getSaveDirectory().append("playersettings.json"));
```
Comment (wlong123) [inline source/CIApp.cpp:77]: Can you put the actual type here? I don't think theres any reason to use auto here b/c you know you're create a JSONWriter

File: source/CIApp.cpp [hunk 5]
```diff
@@ -191,14 +175,11 @@ void CoreImpactApp::update(float timestep) {
  */
 void CoreImpactApp::onSuspend() {
     // save settings file
-    std::shared_ptr<cugl::JsonWriter> _writer = JsonWriter::alloc(Application::getSaveDirectory().append("settings.json"));
-    std::shared_ptr<cugl::JsonValue> _settings = JsonValue::allocObject();
-    _settings->appendValue("PlayerName", _menu.getPlayerName());
-    _settings->appendValue("Volume", _menu.getVolume());
-    _settings->appendValue("MusicOn", _menu.isMusicOn());
-    _settings->appendValue("ParallaxOn", _menu.isParallaxOn());
+    auto _writer = JsonWriter::alloc(Application::getSaveDirectory().append("playersettings.json"));
```
Comment (wlong123) [inline source/CIApp.cpp:178]: Is there a reason you're writing the player settings in onSuspend and onShutdown?

File: source/CICollisionController.cpp [hunk 6]
```diff
@@ -77,6 +77,11 @@ void collisions::checkForCollision(const std::shared_ptr<PlanetModel>& planet, c
                 float force = timestep * 60 * 9.81f *
                     (stardust->getMass() * stardust->getMass()) *
                     planet->getMass() / (distance * distance);
+                
+                // handle game settings 
+                auto pGrav = planet->getGravStrength();
```
Comment (wlong123) [inline source/CICollisionController.cpp:82]: This should always be a float right?

File: source/CIGameScene.cpp [hunk 7]
```diff
@@ -80,7 +91,7 @@ bool GameScene::init(const std::shared_ptr<cugl::AssetManager>& assets,
     _gameUpdateManager = GameUpdateManager::alloc();
     _networkMessageManager = networkMessageManager;
     _networkMessageManager->setGameuUpdateManager(_gameUpdateManager);
-    if (isHost) {
+    if (gameId.empty()) { // Host
```
Comment (wlong123) [inline source/CIGameScene.cpp:94]: This is not a good way to determine whether this player is the host. Player id being 0 means the player is the host, theres a method for this in network message manager

File: source/CIGameScene.cpp [hunk 8]
```diff
@@ -99,7 +110,7 @@ bool GameScene::init(const std::shared_ptr<cugl::AssetManager>& assets,
     _massHUD  = std::dynamic_pointer_cast<scene2::Label>(_assets->get<scene2::SceneNode>("game_hud"));
 
     // Create the planet model
-    _planet = PlanetModel::alloc(dimen.width/2, dimen.height/2, CIColor::getNoneColor(), 3);
+    _planet = PlanetModel::alloc(dimen.width / 2, dimen.height / 2, CIColor::getNoneColor(), MAX_PLANET_LAYERS, gravStrength, winCondition);
```
Comment (wlong123) [inline source/CIGameScene.cpp:113]: Could you change the name winCondition to something a bit more specific. Maybe like winningPlanetMass or planetMassToWin

File: source/CIGameScene.cpp [hunk 9]
```diff
@@ -38,8 +38,17 @@ using namespace std;
 /** Maximum number of stardusts allowed on screen at a time. */
 #define MAX_STARDUST 512
 
-/** Starting value for the countdown */
-#define START_COUNTDOWN -10.0f
+/** Maximum number of layers in planet model */
+#define MAX_PLANET_LAYERS 3
+
+/** Default value for the countdown (inactive) */
+#define INACTIVE_COUNTDOWN -10.0f
```
Comment (wlong123) [inline source/CIGameScene.cpp:45]: Can you name this something more specific. Same with start countdown. Both these variables should say something about winning

File: source/CIGameScene.h [hunk 10]
```diff
@@ -81,7 +87,7 @@ class GameScene : public cugl::Scene2 {
      * This constructor does not allocate any objects or start the game.
      * This allows us to use the object without a heap pointer.
      */
-    GameScene() : cugl::Scene2() {
+    GameScene() : cugl::Scene2(), _spawnRate(40.0f), _colorCount(6), _countdown(-10.0f) {
```
Comment (wlong123) [inline source/CIGameScene.h:90]: You can put these constants in the header file and they'll also be included in the cpp file. i.e. BASE_SPAWN_RATE and INACTIVE_COUNTDOWN should be in this file

File: source/CILobbyMenu.cpp [hunk 11]
```diff
@@ -3,7 +3,7 @@
 //  CoreImpact
 //
 //  Created by Richard Yoon on 4/5/21.
-//  Copyright � 2021 Game Design Initiative at Cornell. All rights reserved.
+//  Copyright � 2021 Game Design Initiative at Cornell. All rights reserved.
```
Comment (wlong123) [inline source/CILobbyMenu.cpp:6]: Why was this changed?

File: source/CIPlanetModel.cpp [hunk 12]
```diff
@@ -54,6 +58,27 @@ void PlanetModel::dispose() {
...
+ * @param x The initial x-coordinate of the center
+ * @param y The initial y-coordinate of the center
+ * @param c The initial color code of the planet
+ * @param maxLayers The maximum number of layers the planet can have
+ * @param gravStrength The planet's gravitational strength factor
+ * @param winCond The condition value for the planet to win
+ *
+ * @return true if the initialization was successful
+ */
+bool PlanetModel::init(float x, float y, CIColor::Value c, int maxLayers, float gravStrength, uint16_t winCond) {
```
Comment (wlong123) [inline source/CIPlanetModel.cpp:76]: Is the other init method still used? If not we should get rid of it or put default values for maxLayers and gravStrength

File: source/CIGameScene.cpp [hunk 13]
```diff
@@ -280,17 +299,17 @@ void GameScene::addStardust(const Size bounds) {
...
-    int probSum = 0, colorCount = 6;
+    int probSum = 0;
     // Sums up the total probability space of the stardust colors, augmented by a mass correction
-    probSum = accumulate(_stardustProb, _stardustProb + colorCount, probSum) + massCorrection;
+    probSum = accumulate(_stardustProb, _stardustProb + _colorCount, probSum) + massCorrection;
     // Randomly selects a point in the probability space
-    int spawnRand = rand() % probSum;
-    for (int i = 0; i < colorCount; i++) {
+    int spawnRand = rand() % max(1, probSum);
+    for (int i = 0; i < _colorCount; i++) {
```
Comment (wlong123) [inline source/CIGameScene.cpp:307]: you should keep the stardust spawning the same as before or update all of the stardust spawning to be consistent with _colorCount.

File: source/CILobbyMenu.cpp (hunk 11 above)
Comment (yds05238) [reply source/CILobbyMenu.cpp:6]: Im not sure where thats from. Ill revert it back to the original line.

File: source/CIApp.cpp (hunk 5 above)
Comment (yds05238) [reply source/CIApp.cpp:178]: The player settings was resetting on every new game instance. Saving the settings on both occasions allows the subsequent game instances to load in the saved settings instead of resetting.

File: source/CIGameScene.cpp (hunk 8 above)
Comment (yds05238) [reply source/CIGameScene.cpp:113]: I wasn't sure if we were planning to switch to using number of layers instead of the planet's mass to determine winning. Ill update the code to go along with using the planet mass

File: source/CIPlanetModel.cpp (hunk 12 above)
Comment (yds05238) [reply source/CIPlanetModel.cpp:76]: I think its being called in OpponentPlanet's alloc method.

File: source/CIGameScene.h (hunk 10 above)
Comment (yds05238) [reply source/CIGameScene.h:90]: Should I create a header file for all the default/constant values in menu scenes? It would essentially be a namespace with all the menu constants defined.

File: source/CIApp.cpp (hunk 5 above)
Comment (wlong123) [reply source/CIApp.cpp:178]: okay I think you can get of the onShutdown saving then? I think?

File: source/CIGameScene.h (hunk 10 above)
Comment (wlong123) [reply source/CIGameScene.h]: I don't think that's necessary b/c there aren't too many constants. We may want a global constants file at some point but that's definitely not needed for this pr

File: source/CIPlanetModel.cpp (hunk 12 above)
Comment (wlong123) [reply source/CIPlanetModel.cpp]: ok can you do the default values thing then. Take a look at addStardust in StardustQueue.h for an example

File: source/CIApp.cpp (hunk 5 above)
Comment (wlong123) [inline source/CIApp.cpp:178]: also add type here

File: source/CIApp.cpp [hunk 14]
```diff
@@ -191,14 +175,11 @@ void CoreImpactApp::update(float timestep) {
...
+    auto _settings = JsonValue::allocObject();
```
Comment (wlong123) [inline source/CIApp.cpp:179]: type

File: source/CIGameScene.cpp [hunk 15]
```diff
@@ -325,16 +344,18 @@ void GameScene::processSpecialStardust(const cugl::Size bounds, const std::share
...
-                stardustQueue->addStardust(CIColor::getRandomColor(), bounds);
-                stardustQueue->addStardust(CIColor::getRandomColor(), bounds);
+                stardustQueue->addStardust(CIColor::getRandomColor(_colorCount), bounds);
+                stardustQueue->addStardust(CIColor::getRandomColor(_colorCount), bounds);
                 break;
             case StardustModel::Type::SHOOTING_STAR:
                 CULog("SHOOTING STAR");
                 stardustQueue->addShootingStardust(stardust->getColor(), bounds);
                 stardustQueue->addShootingStardust(stardust->getColor(), bounds);
+                [[fallthrough]];
```
Comment (wlong123) [inline source/CIGameScene.cpp:354]: you can get rid of the [[failthrough]] here. I put breaks in the pr I just put up

File: source/CIGameScene.h [hunk 16]
```diff
@@ -88,9 +103,10 @@ class GameScene : public cugl::Scene2 {
      * This constructor does not allocate any objects or start the game.
      * This allows us to use the object without a heap pointer.
      */
-    GameScene() : cugl::Scene2() {
+    GameScene() : cugl::Scene2(), _spawnRate(BASE_SPAWN_RATE), _colorCount(DEFAULT_COLOR_COUNTS), _countdown(INACTIVE_WIN_COUNTER) {
         for (int i = 0; i < 6; i++) {
-            _stardustProb[i] = 100;
+            //_stardustProb[i] = 100;
```
Comment (wlong123) [inline source/CIGameScene.h:108]: NIT: remove comment code

**Review by wlong123 — APPROVED** [review]
Overall looks good! Just a few minor things to fix up before merging

File: source/CIColor.h [hunk 17]
```diff
@@ -121,10 +121,10 @@ class CIColor {
      *
      * @return CIColor Value for a stardust
      */
-    static Value getRandomColor() {
+    static Value getRandomColor(int ccount = 5) {
         std::random_device rd;
         std::mt19937 mt(rd());
-        std::uniform_int_distribution<int> dist(0, 5); // pick out of first 6 colors
+        std::uniform_int_distribution<int> dist(0, ccount); // pick out of first 6 colors
```
Comment (steiner26) [inline source/CIColor.h:127]: Can we remove the comment here?

File: source/CIMenuScene.h [hunk 18]
```diff
@@ -80,7 +90,8 @@ class MenuScene : public cugl::Scene2 {
      * This constructor does not allocate any objects or start the game.
      * This allows us to use the object without a heap pointer.
      */
-    MenuScene() : cugl::Scene2(), _volume(0.0f), _musicOn(true), _parallaxOn(true) {}
+    MenuScene() : cugl::Scene2(), _volume(0.0f), _musicOn(true), _parallaxOn(true),
```
Comment (steiner26) [inline source/CIMenuScene.h:93]: I know this will just get overwritten in init, but can we make the initial value for volume 0.5?

File: source/CIApp.cpp (hunk 5 above)
Comment (yds05238) [reply source/CIApp.cpp:178]: So on the iphone, the program does not go through onShutdown due to how the phone handles app shutdown/suspend. The UI forces app to suspend before shutting down which prevents the program from triggering the onShutdown method.

File: source/CIPlanetNode.h [hunk 19]
```diff
@@ -17,7 +17,7 @@
 #include "CIPlanetLayer.h"
 
 #define PLANET_RING_TEXTURE_INNER_SIZE 140
-#define PLANET_OUTER_RING_SCALE 0.833
+#define PLANET_OUTER_RING_SCALE 1.770
```
Comment (steiner26) [inline source/CIPlanetNode.h:20]: Can you change line 96 in PlnaetNode.cpp to be the same as PlanetNode.h? The initial creation of the node outer ring should use this scale the same way that setRadius does

**Review by steiner26 — APPROVED** [review]
Looks good to me, just a 1 changes and some style things to change before its merged in

File: source/CIPlanetNode.h (hunk 19 above)
Comment (yds05238) [reply source/CIPlanetNode.h:20]: So replace the value passed in line 96 with "PLANET_OUTER_RING_SCALE * _layerScale/_coreScale"?
